lib/improver/spotdata/extrema.py

```diff
@@ -0,0 +1,255 @@
+
+    e.g.
+
+    UTC Coord  :  12   13   14   15   16
+    Data: Site 1  300  302  296  294  290 (UTC offset = -2)
+          Site 2  280  282  283  280  279 (UTC offset = +2)
+
+    Data redistributed according to UTC offset to sit on local time.
+
+    Local times:  10   11   12   13   14   15   16   17   18
+    Data: Site 1  300  302  296  294  290  -    -    -    -
+          Site 2  -    -    -    -    280  282  283  280  279
+
+    There will be missing but masked data in locations which are not forecast
+    at the given local time, e.g. a SpotData site at a UTC location for a 12Z
+    run will have no forecast temperatures on the local time axis for times
+    earlier than 12Z.
+
+    Maxima/Minima can then be calculated on local time, as makes sense for
+    quantities such as maximum in day.
+
+    New index is obtained as:
+        Site_UTC_offset + 12 + UTC_coordinate_index
+
+        +12 ensures that UTC offset -12 sits at an index of 0 etc.
+
+        UTC_coordinate_index is the data's index on the original coordinate
+        axis. So imagining a model run with a forecast reference time of 12Z,
+        a temperature at 14Z UTC will sit at index=2. At a SpotData site with
+        a UTC offset of -6, the index will become 8:
+
+   Index: 0  1  2  3  4  5  6  7  8  9  10 11 12
+   UTC    12 13 14 15 16 17 18 19 20 21 22 23 00
+   Data :       X
+   Local: 00 01 02 03 04 05 06 07 08 09 10 11 12
+   Data :                         X
+
+
+    Args:
+    -----
+    cube : iris.cube.Cube
+        A cube of site data resulting from the SpotData extraction process.
+
+    Returns:
+    --------
+    cube : iris.cube.Cube
+        A cube with localised time coordinates.
+
+    """
+    # Ensure time coordinate is in hours.
+    hour_coordinates = cube.coord('time')
+    hour_coordinates.convert_units('hours since 1970-01-01 00:00:00')
+
+    # Calculate local time range which spans earliest time UTC -12 for far west
+    # to latest time UTC +14 in far east. Ignores half hour time zones.
+    local_time_min = int(hour_coordinates.points[0]-12)
+    local_time_max = int(hour_coordinates.points[-1]+14)
+    local_times = range(local_time_min, local_time_max+1, 1)
+
+    # Create iris.coord.DimCoord of local times.
+    local_time_coord = DimCoord(local_times, standard_name='time',
+                                units=hour_coordinates.units)
+
+    # Create empty array to contain extrema data.
+    new_data = np.full((len(local_times), cube.data.shape[1]), np.nan)
+
+    # Create ascending indices to help with filling new_data array.
+    n_sites = cube.data.shape[1]
+    row_index = range(0, n_sites)
+
+    # Loop through times in UTC and displace data in array so that each datum
+    # sits at its local time.
+    for i_time in range(len(hour_coordinates.points)):
+        indices = cube.coord('utc_offset').points.astype(int) + 12 + i_time
+        new_data[indices, row_index] = cube.data[i_time]
```

Comment:
Height is unlikely to be a dimension coordinate as we are always talking surface obs in spotdata (as far as I know). It could certainly be made more generic in future I imagine, but I'd like to suggest we leave that out of the current PR and come back to these functions as and when additional utility is required.